### ISSUE TITLE
`plyranges`: `join_overlap` but also include distance column

### ISSUE THREAD

**Issue by mikelove** [tidyomics/genomics-todos#2]
As a crude example:

https://gist.github.com/mikelove/b855d493656438c0de64901d44b3eafc

Would be nice to allow this behavior within a function e.g. `join_overlap_*_with_distance`

Choices: 5', 3', center, classic distance?

**Comment by jennprk** [comment]
This is actually something I needed for my research. I'd love to make a function for this if it's still needed !

**Comment by mikelove** [comment]
Going to start working here
https://github.com/tidyomics/plyranges/tree/join_with_dist

Guiding ideas:
* only for inner and left join, directed or not, but not for `within`
* add a simple argument `distance=FALSE` so doesn't affect current pipelines
* need to figure out logic behind distance... hopefully not implementing a lot of choices

**Comment by ppaxisa** [comment]
Since group mutate seems to be good to go, maybe I can help here? Or you already have this worked out?

**Comment by mikelove** [comment]
Awesome that would be great if you can help out and thanks for the work on group mutate.

I chatted with @sa-lee this week and he mentioned he may be refactoring overlap joins in the upcoming devel cycle. In which case we should take that into account. 

Maybe we can discuss at an upcoming devel meeting? (What times are good for you?) There are considerations like how to do this without introducing much new code. There are a bunch of join functions but ideally we could have this feature with only small addition in one central place. If the code is refactored this may be easier to accomplish.

**Comment by ppaxisa** [comment]
It should be easy (and compact) with `GenomicRanges::distance` given that in `.join_overlap_left` there is:
```r
# overlaps found
   left <- x[queryHits(hits), ]
   right <- y[subjectHits(hits), ]
```
so we could add `mcols(left)$distance <- distance(left, right)`. `distance` also has a strand argument but I always struggle to reason with strand, not sure about what should the intended behaviour be.

Happy to wait for refactoring too!

**Comment by mikelove** [comment]
Agree, this looks like it could be it! I'll run it by @sa-lee and see if it's relevant for refactoring if if we could add this now. As an interface issue, I think we'd have to expose some argument `add_distance` in multiple functions.

**Comment by sa-lee** [comment]
This seems good to me, most of the join functions, are using the hit mechanisms underneath, so should be fairly straighforward to add distance using the approach @ppaxisa mentioned. 

That being said, I think we could add two options here:
1. Add an `include_distance` or `add_distance` argument to all existing join functions, where relevant.
2. Add a function called `distance_to()` that could be evaluated inside things like mutate etc. That computes the distance from current range to target range or vice verse. 

I was thinking of refactoring this api somewhat to align it with [`dplyr::join_by()`](https://dplyr.tidyverse.org/reference/join_by.html) but I don't think that's a seperate issue.

**Comment by mikelove** [comment]
I'm mostly interested in number 1 here, because i want metadata from the target range as well as the distance.

I may take a look at this in the next few weeks unless someone already has code.

**Comment by sa-lee** [comment]
happy to collaborate with you on this too

**Comment by mikelove** [comment]
Great. Let me give a shot and send off to you for review

**Comment by mikelove** [comment]
Starting this here:

https://github.com/tidyomics/plyranges/blob/join_with_dist/R/ranges-overlap-joins-outer.R#L63-L67

I plan to add a directed version (passing `ignore.strand` down), and I guess repeat everything for `_inner`.

**Comment by mikelove** [comment]
@sa-lee if you have a chance, this would be a very basic addition. I would do this also with respect to `ignore.strand` and `*_inner`. 

*_within* is not needed as this is for use when ranges are *not* within.

Tests here https://github.com/tidyomics/plyranges/blob/86da4a7835505d794161c469e406a8c4c868c723/tests/testthat/test-ojoins-distance.R

**Comment by mikelove** [comment]
This is folded into the main devel branch on github.

Current implemented for `join_overlap_left` and `_directed` variant, but could be extended to others easily. I need to check that I didn't break anything with the IRanges method.
